### PR TITLE
Dialogue Unlikelihood OS commit

### PR DESCRIPTION
**Pull request by hadasah**
**Patch description**
Open sourcing our dialogue unlikelihood code

**Testing steps**
`pytest -v -s tests/nightly/gpu/test_unlikelihood.py`

**Logs**
```

```
Currently the ELI5 tests are failing because the ELI5 task isn't built the way the rest of our tasks are.

### CODE REVIEW

File: parlai/core/build_data.py [hunk 1]
```diff
@@ -383,7 +383,7 @@ def download_models(
                     url += model_type + '/'
                 url += fname
             else:
-                url = path + '/' + fname
+                url = path + '/' + model_folder + '/' + fname
```
Comment (emilydinan) [inline parlai/core/build_data.py:386]: hmmm... this worries me. will this break anything?

File: parlai/agents/image_seq2seq/image_seq2seq.py [hunk 2]
```diff
@@ -55,6 +55,7 @@ def add_cmdline_args(cls, argparser):
             recommended=True,
             help='if true, include image token (or no image token) for each example',
         )
+        return group
```
Comment (emilydinan) [inline parlai/agents/image_seq2seq/image_seq2seq.py:58]: does this change belong here?

File: projects/dialogue_unlikelihood/agents.py [hunk 3]
```diff
@@ -0,0 +1,619 @@
...
+
+        maxlen = self.label_truncate or 256
+        beam_preds_scores, _ = self._generate(batch, self.beam_size, maxlen)
+        preds, scores = zip(*beam_preds_scores)
+
+        self._ngram_metrics(batch, preds)
+
+        cand_choices = None
+        # TODO: abstract out the scoring here?
+        if self.rank_candidates:
```
Comment (emilydinan) [inline projects/dialogue_unlikelihood/agents.py:276]: would be nice if this was abstracted out in TGA so that we avoid the copypasta here

File: projects/dialogue_unlikelihood/agents.py [hunk 4]
```diff
@@ -0,0 +1,619 @@
...
+    """
+
+    pass
+
+
+class RepetitionUnlikelihoodAgent(RepetitionUnlikelihoodAgentTrait, ImageSeq2seqAgent):
+    """
+    Example usage:
+
+        -t internal:generation_safety -m parlai_internal.projects.dontsaythat.agents:RepetitionUnlikelihoodAgent
```
Comment (emilydinan) [inline projects/dialogue_unlikelihood/agents.py:616]: have to get rid of reference to parlai_internal here and elsewhere

File: parlai/core/build_data.py (hunk 1 above)
Comment (hadasah) [reply parlai/core/build_data.py:386]: ah, I can get rid of this. I was using this during localhost testing (if you follow the quip we have, you don't end up wget-ing the right file)

File: parlai/agents/image_seq2seq/image_seq2seq.py (hunk 2 above)
Comment (hadasah) [reply parlai/agents/image_seq2seq/image_seq2seq.py:58]: yes, my repetitionunlikelihood agent inherits from image_seq2seq because we used an image_seq2seq model as our ELI5 baseline. when I call super().add_cmdline_args() from there, I get None without this line

**Comment by stephenroller** [comment]
Hmm, looks like we have issues since we can't distribute the eli5 data. Let's think about alternatives to this.

File: projects/dialogue_unlikelihood/agents.py [hunk 5]
```diff
@@ -0,0 +1,619 @@
...
+
+        if batch.label_vec is not None:
+            # calculate loss on targets with teacher forcing
+            _ = self.compute_loss(batch)  # noqa: F841  we need the side effects
+
+        maxlen = self.label_truncate or 256
+        beam_preds_scores, _ = self._generate(batch, self.beam_size, maxlen)
+        preds, scores = zip(*beam_preds_scores)
+
+        self._ngram_metrics(batch, preds)
```
Comment (stephenroller) [inline projects/dialogue_unlikelihood/agents.py:272]: all of this copy pasta is just so we can run this method, isn't it? Can we just carve out something in TGA that we can override?

**Comment by stephenroller** [comment]
Margaret can you rebase this and fix the lint errors?

**Comment by stephenroller** [comment]
So go ahead and throw a @unittest.skip on the ELI5 tests, and then fix the main unit tests breaks, and I think this is ready.

File: parlai/core/build_data.py (hunk 1 above)
Comment (stephenroller) [reply parlai/core/build_data.py:386]: drop this still tho

File: projects/dialogue_unlikelihood/agents.py (hunk 5 above)
Comment (stephenroller) [reply projects/dialogue_unlikelihood/agents.py:276]: it still would be kinda nice to see this not be the case. idk.

Comment (hadasah) [reply projects/dialogue_unlikelihood/agents.py:242]: I added a call to a new method _add_generation_metrics() in TGA's eval_step(), which I implement in the unlikelihood agent. Is this good or were you thinking of something different?

Comment (stephenroller) [reply projects/dialogue_unlikelihood/agents.py:242]: That's awesome.

Comment (stephenroller) [reply projects/dialogue_unlikelihood/agents.py:242]: i don't think the add to TGA's eval_step made it in to the PR, make sure you add that.

**Comment by stephenroller** [comment]
🎉🎉🎉
